File: internal/dbprobe/dbprobe.go
// Package dbprobe runs the receiver's monitoring SQL directly against Oracle
// (ground truth), using the same driver the receiver uses: sijms/go-ora.
//
// go-ora is a pure-Go driver that negotiates Oracle Native Network Encryption,
// so unlike the python-oracledb thin client this needs no Oracle Instant Client
// and no "thick mode".
package dbprobe

import (
	"database/sql"
	"fmt"
	"strings"
	"time"

	go_ora "github.com/sijms/go-ora/v2"

	"github.com/newrelic-forks/otel-metric-validator/internal/config"
	"github.com/newrelic-forks/otel-metric-validator/internal/metricmap"
)

// Result holds the expected data points derived from the DB and the probe time.
type Result struct {
	Expected  map[string]metricmap.Expected
	ProbeTime time.Time
	Errors    []string
}

// DSN builds the go-ora connection URL exactly as the receiver does.
func DSN(cfg config.Config) string {
	return go_ora.BuildUrl(cfg.Host, cfg.Port, cfg.Service, cfg.User, cfg.Password, nil)
}

// Probe connects, runs every mapped query once, and returns expected data points.
func Probe(cfg config.Config) (Result, error) {
	db, err := sql.Open("oracle", DSN(cfg))
	if err != nil {
		return Result{}, fmt.Errorf("open: %w", err)
	}
	defer db.Close()

	res := Result{Expected: map[string]metricmap.Expected{}, ProbeTime: time.Now()}
	for _, key := range metricmap.AllQueryKeys() {
		query := metricmap.QuerySQL(key, cfg.IsCDB())
		rows, err := runQuery(db, query)
		if err != nil {
			res.Errors = append(res.Errors, fmt.Sprintf("query %q failed: %v", key, err))
			continue
		}
		for _, exp := range metricmap.ExpectedFor(key, rows) {
			res.Expected[exp.Key()] = exp
		}
	}
	return res, nil
}

// runQuery executes SQL and returns rows as maps with UPPERCASE column names.
// Duplicate column names (e.g. the resource-limit LIMIT_VALUE) follow last-wins,
// matching the receiver's row map.
func runQuery(db *sql.DB, query string) ([]map[string]any, error) {
	rows, err := db.Query(query)
	if err != nil {
		return nil, err
	}
	defer rows.Close()

	cols, err := rows.Columns()
	if err != nil {
		return nil, err
	}
	upper := make([]string, len(cols))
	for i, c := range cols {
		upper[i] = strings.ToUpper(c)
	}

	var out []map[string]any
	for rows.Next() {
		holders := make([]any, len(cols))
		ptrs := make([]any, len(cols))
		for i := range holders {
			ptrs[i] = &holders[i]
		}
		if err := rows.Scan(ptrs...); err != nil {
			return nil, err
		}
		row := make(map[string]any, len(cols))
		for i, name := range upper {
			row[name] = holders[i] // later duplicate columns overwrite earlier ones
		}
		out = append(out, row)
	}
	return out, rows.Err()
}

File: validator/db_probe.py
"""Run the receiver's monitoring SQL directly against Oracle (ground truth).

Uses python-oracledb in thin mode by default (no Oracle Instant Client needed).
If the DB server enforces Native Network Encryption (NNE) — thin mode cannot
negotiate it — set VALIDATOR_ORACLE_THICK=1 (and optionally ORACLE_CLIENT_LIB_DIR)
to switch to thick mode, which requires an Oracle Instant Client install.
"""

from __future__ import annotations

import time
from dataclasses import dataclass, field

from . import metric_map
from .config import Config

_thick_initialized = False


def _ensure_thick_mode(oracledb, cfg: Config) -> None:
    """Initialize the Oracle Instant Client once, for NNE-enforcing servers."""
    global _thick_initialized
    if _thick_initialized:
        return
    kwargs = {"lib_dir": cfg.oracle_client_lib_dir} if cfg.oracle_client_lib_dir else {}
    oracledb.init_oracle_client(**kwargs)
    _thick_initialized = True


@dataclass
class ProbeResult:
    expected: dict[tuple, metric_map.Expected]  # Expected.key() -> Expected
    probe_time: float                            # epoch seconds when the probe ran
    errors: list[str] = field(default_factory=list)


def _rows_as_dicts(cursor) -> list[dict]:
    """Fetch all rows as dicts with UPPERCASE column names.

    Oracle returns duplicate column names for the resource-limit query (raw and
    normalized LIMIT_VALUE); last column wins, matching the receiver's row map.
    """
    cols = [d[0].upper() for d in cursor.description]
    result = []
    for rec in cursor.fetchall():
        row: dict[str, object] = {}
        for col, val in zip(cols, rec):
            row[col] = val  # later duplicate columns overwrite earlier ones
        result.append(row)
    return result


def probe(cfg: Config) -> ProbeResult:
    """Connect, run every mapped query once, and return expected data points."""
    import oracledb  # imported here so unit tests need not have the driver installed

    expected: dict[tuple, metric_map.Expected] = {}
    errors: list[str] = []

    if cfg.oracle_thick:
        _ensure_thick_mode(oracledb, cfg)
    conn = oracledb.connect(user=cfg.user, password=cfg.password, dsn=cfg.dsn)
    probe_time = time.time()
    try:
        for key in metric_map.all_query_keys():
            sql = metric_map.query_sql(key, cfg.is_cdb)
            try:
                with conn.cursor() as cur:
                    cur.execute(sql)
                    rows = _rows_as_dicts(cur)
            except Exception as exc:  # noqa: BLE001 - surface per-query failures, keep going
                errors.append(f"query {key!r} failed: {exc}")
                continue
            for exp in metric_map.expected_for(key, rows):
                expected[exp.key()] = exp
    finally:
        conn.close()

    return ProbeResult(expected=expected, probe_time=probe_time, errors=errors)
